### PR TITLE
Finetune model & Extract feature 

### PR DESCRIPTION
**Pull request by antinucleon**

### CODE REVIEW

File: python/mxnet/model.py [hunk 1]
```diff
@@ -916,3 +917,89 @@ def create(symbol, X, y=None, ctx=None,
...
+    -------
+    inputs: list
+        A list of input node info
+    """
+    inputs = []
+    for node_idx, output_idx in node_dict["inputs"]:
+        if node_idx in arg_dict:
+            continue
+        else:
+            inputs.append([node_idx, output_idx])
```
Comment (haijieg) [inline python/mxnet/model.py]: This  might be cleaner:

```
if node_idx not in arg_dict:
    inputs.append([node_idx, output_idx])
```

File: python/mxnet/model.py [hunk 2]
```diff
@@ -916,3 +917,89 @@ def create(symbol, X, y=None, ctx=None,
...
+    return inputs
+
+def _get_feature_symbol(sym_json):
+    """Get feature extractor symbol from the given symbol
+    Note: This function is NOT working with multiple output network
```
Comment (haijieg) [inline python/mxnet/model.py]:
```
.. note::
    This function does not work ...
```

File: python/mxnet/model.py [hunk 3]
```diff
@@ -916,3 +917,89 @@ def create(symbol, X, y=None, ctx=None,
...
+    """
+
+    sym_dict = json.loads(sym_json)
+    arg_nodes = sym_dict["arg_nodes"]
+    arg_dict = dict(zip(arg_nodes, [True for i in range(len(arg_nodes))]))
+    heads = sym_dict["heads"]
+    nodes = sym_dict["nodes"]
+
+    if len(heads) != 1:
+        raise Exception("Only support sinle output model")
```
Comment (haijieg) [inline python/mxnet/model.py]: sinle -> single

File: python/mxnet/model.py [hunk 4]
```diff
@@ -916,3 +917,89 @@ def create(symbol, X, y=None, ctx=None,
...
+    head_idx = heads[0][0]
+    # get linear classifier node
+    cls = _find_nodes_inputs(nodes[head_idx], arg_dict)
+    if len(cls) != 1:
+        raise Exception("Only support sinle output model")
```
Comment (haijieg) [inline python/mxnet/model.py]: sinle -> single

File: python/mxnet/model.py [hunk 5]
```diff
@@ -916,3 +917,89 @@ def create(symbol, X, y=None, ctx=None,
...
+    sym_dict = json.loads(sym_json)
+    arg_nodes = sym_dict["arg_nodes"]
+    arg_dict = dict(zip(arg_nodes, [True for i in range(len(arg_nodes))]))
+    heads = sym_dict["heads"]
+    nodes = sym_dict["nodes"]
+
+    if len(heads) != 1:
+        raise Exception("Only support sinle output model")
+    head_idx = heads[0][0]
+    # get linear classifier node
```
Comment (haijieg) [inline python/mxnet/model.py]: what does the comment mean here?

File: python/mxnet/model.py [hunk 6]
```diff
@@ -916,3 +917,89 @@ def create(symbol, X, y=None, ctx=None,
...
+    cls = _find_nodes_inputs(nodes[head_idx], arg_dict)
```
Comment (haijieg) [inline python/mxnet/model.py]: cls is a reserved keywork in python.

File: python/mxnet/model.py [hunk 7]
```diff
@@ -916,3 +917,89 @@ def create(symbol, X, y=None, ctx=None,
...
+
+    # make new symbol
+    new_nodes = nodes[:fea[0][0] + 1]
+    new_arg_nodes = [idx for idx in arg_nodes if idx < fea[0][0]]
+    new_sym = {"arg_nodes" : new_arg_nodes, "heads" : fea, "nodes":new_nodes}
+    return sym.load_json(json.dumps(new_sym))
+
+def finetune(model, num_new_output, **kwargs):
+    """Get a FeedForward model for fine-tune
+    Note: This function is not working with multiple outputs network
```
Comment (haijieg) [inline python/mxnet/model.py]: Use 

```
.. note::

    stuff stuff
```

See http://www.sphinx-doc.org/en/stable/markup/para.html

File: python/mxnet/model.py [hunk 8]
```diff
@@ -916,3 +917,89 @@ def create(symbol, X, y=None, ctx=None,
...
+        Model which contains parameters which will be used for fine-tune
+    num_new_output: int
+        New number of output category
+    kwargs: kwargs
+        mx.model.create function's parameter
+    """
+    sym_json = model.symbol.tojson()
+    fea = _get_feature_symbol(sym_json)
+    cls = sym.FullyConnected(fea, num_hidden=num_new_output)
+    cls = sym.SoftmaxOutput(cls, name="softmax")
```
Comment (haijieg) [inline python/mxnet/model.py]: Avoid using variable named `cls`.
The last output layer may not always be a softmax output. Should we infer the output layer from the input model?

Comment (antinucleon) [inline python/mxnet/model.py]: after we have new model, we can use `list_internals` to get other outputs

File: python/mxnet/model.py [hunk 9]
```diff
@@ -916,3 +917,89 @@ def create(symbol, X, y=None, ctx=None,
...
+        Model which contains parameters which will be used for fine-tune
+    num_new_output: int
+        New number of output category
+    kwargs: kwargs
+        mx.model.create function's parameter
+    """
+    sym_json = model.symbol.tojson()
+    feature = _get_feature_symbol(sym_json)
+    classifier = sym.FullyConnected(feature, num_hidden=num_new_output)
+    classifier = sym.SoftmaxOutput(classifier, name="softmax")
```
Comment (haijieg) [inline python/mxnet/model.py]: Here can we use the same output layer as the original network, i.e. (same type, same name)? This will make finetune work for Regression, BinaryClassification and MulticlassClassification.

**Comment by haijieg** [comment]
Can we add a unittest?

**Comment by antinucleon** [comment]
OK.

File: python/mxnet/model.py [hunk 10]
```diff
@@ -916,3 +917,87 @@ def create(symbol, X, y=None, ctx=None,
...
+    new_model: mx.model.FeedForward
+        Finetuned model
+    """
+    initializer = Load(param=model.arg_params, default_init=Uniform(0.001))
+    new_model = FeedForward.create(symbol=symbol, initializer=initializer, **kwargs)
+    return new_model
+
+
+def extractfeature(model, data, top_layer=None, **kwargs):
+    """Extract feature from model
```
Comment (piotr-teterwak) [inline python/mxnet/model.py:997]: Semantics, but I kind of like `extract_feautres` for consistency with GLC.

File: python/mxnet/model.py [hunk 11]
```diff
@@ -916,3 +917,87 @@ def create(symbol, X, y=None, ctx=None,
...
+    top_layer: str, optional
+        Layer which will be extracted. If not set, will extract the second last layer
+    kwargs: kwargs
+        mx.model.FeedForward class' parameters
+
+    Returns:
+    --------
+    pred: numpy.ndarray
+        Extracted feature
+
```
Comment (piotr-teterwak) [inline python/mxnet/model.py:1014]: A small example would be awesome.

File: python/mxnet/model.py [hunk 12]
```diff
@@ -916,3 +917,87 @@ def create(symbol, X, y=None, ctx=None,
...
+    Parameters
+    ----------
+    symbol: mx.symbol.Symbol
+        Symbol of new network will be finetuned.
+    model: mx.model.FeedForward
+        Model which contains parameters which will be used for fine-tune.
+    kwargs: kwargs
+        mx.model.create function's parameters
+
```
Comment (piotr-teterwak) [inline python/mxnet/model.py:967]: Again, small example here would be awesome.

File: python/mxnet/model.py [hunk 13]
```diff
@@ -916,3 +917,87 @@ def create(symbol, X, y=None, ctx=None,
                   work_load_list=work_load_list,
                   eval_batch_end_callback=eval_batch_end_callback)
         return model
+
+def get_feature_symbol(model, top_layer=None):
```
Comment (piotr-teterwak) [inline python/mxnet/model.py:921]: Should this be hidden from top level with an underscore, maybe? Not sure.

File: python/mxnet/model.py [hunk 14]
```diff
@@ -916,3 +917,117 @@ def create(symbol, X, y=None, ctx=None,
...
+        if top_layer not in outputs:
+            error_msg = "%s not exists in symbol. Possible choice:\n%s" \
+                    % (top_layer, "\n".join(outputs))
+            raise ValueError(error_msg)
+    return internals[top_layer]
+
+def finetune(symbol, model, **kwargs):
+    """Get a FeedForward model for fine-tune
+    .. note::
+        For layer doesn't exist in model, will be initialized as uniform random weight
```
Comment (chrisdubois) [inline python/mxnet/model.py:957]: -> Any layers that do not exist in the model will be randomly initialized with values drawn uniformly from [-1, 1]. (or whatever range)

File: python/mxnet/model.py [hunk 15]
```diff
@@ -916,3 +917,117 @@ def create(symbol, X, y=None, ctx=None,
...
+    Load a model
+
+    >>> feature = mx.model.get_feature_symbol(model)
+    >>> net = mx.sym.FullyConnected(data=feature, num_hidden=10, name="new_fc")
+    >>> net = mx.sym.SoftmaxOutput(data=net, name="softmax")
+
+    Make new symbol for finetune
+
+    >>> new_model = mx.model.finetune(symbol=net, model=model, num_epoch=2, learning_rate=1e-3,
+                                      X=train, eval_data=val,
```
Comment (chrisdubois) [inline python/mxnet/model.py:987]: Is there a way I could ask this `net` object what format `train` needs to have? Won't there be some `shape` that id determined by `feature`?

File: python/mxnet/model.py [hunk 16]
```diff
@@ -916,3 +917,117 @@ def create(symbol, X, y=None, ctx=None,
...
+                                      batch_end_callback=mx.callback.Speedometer(100))
+
+    Finetune the model
```
Comment (chrisdubois) [inline python/mxnet/model.py]: This text should be above `>>> new_model...`, no?

File: python/mxnet/model.py [hunk 17]
```diff
@@ -916,3 +917,117 @@ def create(symbol, X, y=None, ctx=None,
...
+    Returns
+    -------
+    new_model: mx.model.FeedForward
+        Finetuned model
+
+    Examples
+    --------
+    >>> sym, arg_params, aux_params = mx.model.load_checkpoint("model", 9)
+
+    Load a model
```
Comment (chrisdubois) [inline python/mxnet/model.py]: This text should be above `>>> sym...`, no?

File: python/mxnet/model.py [hunk 18]
```diff
@@ -916,3 +917,117 @@ def create(symbol, X, y=None, ctx=None,
...
+
+    >>> feature = mx.model.get_feature_symbol(model)
+    >>> net = mx.sym.FullyConnected(data=feature, num_hidden=10, name="new_fc")
+    >>> net = mx.sym.SoftmaxOutput(data=net, name="softmax")
+
+    Make new symbol for finetune
```
Comment (chrisdubois) [inline python/mxnet/model.py]: This text should be previous to the above codeblock, right?

File: python/mxnet/model.py [hunk 19]
```diff
@@ -916,3 +917,117 @@ def create(symbol, X, y=None, ctx=None,
...
+        mx.model.create function's parameters
+
+    Returns
+    -------
+    new_model: mx.model.FeedForward
+        Finetuned model
+
+    Examples
+    --------
+    >>> sym, arg_params, aux_params = mx.model.load_checkpoint("model", 9)
```
Comment (chrisdubois) [inline python/mxnet/model.py:976]: All the text in this example should probably be comments, that way it's copy-and-paste-able. Also, it'd be nice if we had some examples models (and data) on S3 so that this full example is copy-and-paste-able.

File: python/mxnet/model.py (hunk 15 above)
Comment (antinucleon) [inline python/mxnet/model.py:987]: It 's only symbolic parts, not related to shape. After we have symbol, we can do same thing by calling create

File: python/mxnet/model.py (hunk 18 above)
Comment (antinucleon) [inline python/mxnet/model.py]: fixed
